### PR TITLE
pick node with filter applied to it as entry point

### PR DESCRIPTION
**Pull request by swilly22**

### CODE REVIEW

File: src/arithmetic/algebraic_expression.c [hunk 1]
```diff
@@ -27,6 +27,7 @@ AlgebraicExpressionResult *_AlgebraicExpressionResult_New(GrB_Matrix M, Algebrai
 AlgebraicExpression *_AE_MUL(size_t operand_count) {
     AlgebraicExpression *ae = malloc(sizeof(AlgebraicExpression));
     ae->op = AL_EXP_MUL;
+    ae->operand_cap = operand_count + 2;
     ae->operand_count = operand_count;
     ae->operands = malloc(sizeof(AlgebraicExpressionOperand) * operand_count);
```
Comment (jeffreylovitz) [inline src/arithmetic/algebraic_expression.c]: Should multiply by `operand_cap`, not count.

File: src/arithmetic/algebraic_expression.h [hunk 2]
```diff
@@ -40,6 +40,7 @@ typedef struct  {
 typedef struct {
     AL_EXP_OP op;                           // Operation to perform.
     size_t operand_count;                   // Number of operands.
+    size_t operand_cap;                     // Number of operands.
```
Comment (jeffreylovitz) [inline src/arithmetic/algebraic_expression.h]: Comment should be something like `Allocated number of operands`

File: src/execution_plan/execution_plan.c [hunk 3]
```diff
@@ -270,19 +271,36 @@ ExecutionPlan* NewExecutionPlan(RedisModuleCtx *ctx, Graph *g,
                 size_t expCount = 0;
                 AlgebraicExpression **exps = AlgebraicExpression_From_Query(ast, pattern, q, &expCount);
                 
-                TRAVERSE_ORDER order = determineTraverseOrder(q, execution_plan->filter_tree, exps, expCount);
+                TRAVERSE_ORDER order = determineTraverseOrder(q, filter_tree, exps, expCount);
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: In the event that neither `src` nor `dest` are filtered, I wonder if `determineTraverseOrder` should check to see if one has a label while the other doesn't (we'd of course prefer to make a `LabelScan` over an `AllNodeScan`). Thoughts?

File: src/execution_plan/optimizations/select_entry_point.c [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    for(int i = 0; i < Vector_Size(aliases); i++) {
+        char *alias;
+        Vector_Get(aliases, i, &alias);
+        // Source is already filtered
+        bool equal = !strcmp(alias, srcAlias);
+        if(equal) break;
+        
+        // See if dest is filtered.
+        destFiltered |= (strcmp(alias, destAlias) == 0);
```
Comment (jeffreylovitz) [inline src/execution_plan/optimizations/select_entry_point.c]: Minor point - we only need to `strcmp` if `destFiltered == false`.

File: src/execution_plan/optimizations/select_entry_point.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    /* If we're here that means source is not filtered
+     * see if dest has a filter applied to it. */
```
Comment (jeffreylovitz) [inline src/execution_plan/optimizations/select_entry_point.c]: If source is filtered, we break out of the above loop, but still reach this point.

Comment (swilly22) [reply src/execution_plan/optimizations/select_entry_point.c]: Great catch!

File: src/execution_plan/ops/op_conditional_traverse.c [hunk 6]
```diff
@@ -44,12 +65,7 @@ OpResult CondTraverseConsume(OpBase *opBase, QueryGraph* graph) {
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_conditional_traverse.c]: I think that `CondTraverse` ops are guaranteed to have only 1 child, but if that is not always true then we'll only properly traverse the first.

File: src/execution_plan/optimizations/reduce_scans.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    }
+}
+
+void reduceScans(ExecutionPlan *plan) {
+    /* Do not try to remove scan operations 
+     * if query is limited, in such cases keeping scan operations
+     * will speed up query processing times, as we'll be able to 
+     * return as early as possible. */
+    if(ResultSet_Limited(plan->result_set))
+        return;
```
Comment (jeffreylovitz) [inline src/execution_plan/optimizations/reduce_scans.c]: Not necessary now, but at some point we should consider some more factors when deciding whether to apply this optimization. Certain aggregate functions and ORDER clauses will require all operations to be completed regardless of what the LIMIT clause specifies.

File: src/execution_plan/optimizations/traverse_order.c [hunk 8]
```diff
@@ -18,19 +18,23 @@ TRAVERSE_ORDER determineTraverseOrder(const QueryGraph *qg,
...
-
     TRAVERSE_ORDER order = TRAVERSE_ORDER_FIRST;
+    
     Vector *aliases = FilterTree_CollectAliases(filterTree);
     size_t aliasesCount = Vector_Size(aliases);
 
     // See if there's a filter applied to the first expression.
     exp = exps[0];
+    firstExpLabeled |= (*exp->src_node)->label != NULL;
+    firstExpLabeled |= (*exp->dest_node)->label != NULL;
```
Comment (jeffreylovitz) [inline src/execution_plan/optimizations/traverse_order.c]: Functionally identical alternate writing:
`bool firstExpLabeled = (*exp->src_node)->label || (*exp->dest_node)->label;`

I don't have a preference on which version is more readable, just putting this out there!

File: src/execution_plan/ops/op_conditional_traverse.c (hunk 6 above)
Comment (swilly22) [reply src/execution_plan/ops/op_conditional_traverse.c]: guaranteed, the only operation which has more than one child is our Join operation.

File: src/execution_plan/optimizations/reduce_scans.c (hunk 7 above)
Comment (swilly22) [reply src/execution_plan/optimizations/reduce_scans.c]: That's a valid point.

**Pull request merged**